File: algorithm/12주차/창섭/No6603_로또.java
package codeBaekJoon._2021;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class No6603_로또 {

    static int n;
    static boolean visited[];
    static int arr[];
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String tc;
        StringTokenizer st;
        while(!(tc=br.readLine()).equals("0")){
            st = new StringTokenizer(tc, " ");
            n = Integer.parseInt(st.nextToken());
            arr = new int[n];
            visited = new boolean[n];
            for(int i=0; i<n ; i++){
                arr[i] = Integer.parseInt(st.nextToken());
            }
            dfs(0, 0);
            sb.append("\n");
        }
        System.out.println(sb.toString());
    }

    private static void dfs(int deep, int start) {
        if(deep==6){
            for (int i = 0; i <n ; i++) {
                if(visited[i]){
                    sb.append(arr[i]+" ");
                }
            }
            sb.append("\n");
            return;
        }
        for (int i = start; i <n; i++) {
            if(!visited[i]){
                visited[i] = true;
                dfs(deep+1, i+1);
                visited[i] = false;
            }
        }
    }
}

File: algorithm/24주차/창섭/No6603_로또.kt
package codeBaekJoon._2021

import java.io.BufferedReader
import java.io.InputStreamReader

/*
 * 문제 풀이 시작: 2021/06/12 4:14 오후
 * 맞은 시각: 2021/06/12 
 * 소요 시간: 
 * @author ventulus95
*/
var sb = StringBuilder()

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    var line = readLine()
    while (!line.equals("0")){
        var st = line.split(" ").map { it.toInt() }
        st = st.subList(1, st.size);
        dfs(6, 0, st, mutableListOf<Int>())
        sb.append("\n")
        line = readLine()
    }
}

fun dfs(deep: Int, curr: Int, st: List<Int>, list: MutableList<Int>) {
    if(deep==0) {
        for (e in list)
            sb.append("$e ")
        sb.append("\n")
    }
    for(i in curr until st.size){
        list.add(st[i])
        dfs(deep-1, i+1, st, list)
        list.remove(st[i])
    }
}
